### PR TITLE
Suggest returning closure as `impl Fn`

### LINKED ISSUE #100936: For function returning simple closure, suggest possible way to specify the return type
Given the following code: [link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=9337a96600130af288cdbc9cb19c9dab)

```rust
fn func() {
    || 123
}

fn main() {
    println!("{}", func()())
}
```

The current output is:

```
Compiling playground v0.0.1 (/playground)
error[[E0308]](https://doc.rust-lang.org/nightly/error-index.html#E0308): mismatched types
 --> src/main.rs:2:5
  |
1 | fn func() {
  |           - help: a return type might be missing here: `-> _`
2 |     || 123
  |     ^^^^^^ expected `()`, found closure
  |
  = note: expected unit type `()`
               found closure `[closure@src/main.rs:2:5: 2:7]`

error[[E0618]](https://doc.rust-lang.org/nightly/error-index.html#E0618): expected function, found `()`
 --> src/main.rs:6:20
  |
1 | fn func() {
  | --------- `func` defined here returns `()`
...
6 |     println!("{}", func()())
  |                    ^^^^^^--
  |                    |
  |                    call expression requires function

Some errors have detailed explanations: E0308, E0618.
For more information about an error, try `rustc --explain E0308`.
error: could not compile `playground` due to 2 previous errors
```

Ideally the output should suggest how to specify the return type when a function returns a closure.

### PR DESCRIPTION
**Pull request by compiler-errors**
Fixes #100936

### CODE REVIEW

File: compiler/rustc_middle/src/ty/print/pretty.rs [hunk 1]
```diff
@@ -1536,6 +1536,34 @@ pub trait PrettyPrinter<'tcx>:
...
+            for (i, arg) in sig.inputs()[0].tuple_fields().iter().enumerate() {
+                if i > 0 {
+                    p!(", ");
+                }
+                p!(print(arg));
+            }
+            p!(")");
+
+            if !sig.output().is_unit() {
+                p!(" -> ", print(sig.output()));
```
Comment (cjgillot) [inline compiler/rustc_middle/src/ty/print/pretty.rs:1561]: What happens if one of the printed types is an APIT?

Comment (compiler-errors) [reply compiler/rustc_middle/src/ty/print/pretty.rs:1561]: I don't think APITs are `is_suggestable`. I can try tho

Comment (compiler-errors) [reply compiler/rustc_middle/src/ty/print/pretty.rs:1561]: Confirmed that this falls back to the old suggestion and doesn't suggest something like `-> Fn() -> impl Fn()` or anything like that. Added test for it.

**Pull request merged**